src/Workspaces/Remote/ServiceHub/Host/RemoteWorkspace.cs

```diff
@@ -116,189 +95,92 @@ public AssetProvider CreateAssetProvider(Checksum solutionChecksum, SolutionAsse
-            AssetProvider assetProvider,
-            Checksum solutionChecksum,
-            int workspaceVersion,
-            bool fromPrimaryBranch,
-            Func<Solution, ValueTask<T>> doWorkAsync,
-            CancellationToken cancellationToken)
-        {
-            // See if anyone else is computing this solution for this checksum.  If so, just piggy-back on that.  No
-            // need for us to force the same computation to happen ourselves.
-            var currentSolution = this.CurrentSolution;
-
-            // We use a reference-counted solution that implements IAsyncDisposable. The computation of 'newSolution'
-            // uses eager cancellation, but the asynchronous disposable applies lazy cancellation to the final task that
-            // causes cancellation to propagate to the backing lazy operation.
-            var refCountedLazySolution = await GetLazySolutionAsync().ConfigureAwait(false);
-            await using var configuredAsyncDisposable = refCountedLazySolution.ConfigureAwait(false);
+            Contract.ThrowIfNull(solutionChecksum);
+            Contract.ThrowIfTrue(solutionChecksum == Checksum.Null);
 
-            // Actually get the solution, computing it ourselves, or getting the result that another caller was
-            // computing. In the event of cancellation, we do not wait here for the refCountedLazySolution to clean up,
-            // even if this was the last use of this solution.
-            var newSolution = await refCountedLazySolution.Target.Task.WithCancellation(cancellationToken).ConfigureAwait(false);
+            var computeDisconnectedSolutionAsync = (CancellationToken cancellationToken) => ComputeDisconnectedSolutionAsync(assetProvider, solutionChecksum, cancellationToken);
 
-            // We may have just done a lot of work to determine the up to date primary branch solution.  See if we
-            // can move the workspace forward to that solution snapshot.
-            if (fromPrimaryBranch)
-                (newSolution, _) = await this.TryUpdateWorkspaceCurrentSolutionAsync(workspaceVersion, newSolution, cancellationToken).ConfigureAwait(false);
+            Func<Solution, CancellationToken, Task<Solution>>? updatePrimaryBranchAsync = updatePrimaryBranch
+                ? (anyBranchSolution, cancellationToken) => this.TryUpdateWorkspaceCurrentSolutionAsync(workspaceVersion, anyBranchSolution, cancellationToken)
+                : null;
 
-            // Store this around so that if another call comes through, they will see the solution we just computed.
-            await SetLastRequestedSolutionAsync(newSolution).ConfigureAwait(false);
+            // Gets or creates a solution corresponding to the requested checksum.  This will always succeed, and will
+            // increment the in-flight of that solution until we decrement it at the end of our try/finally block.
+            var solution = await GetOrCreateSolutionAndAddInFlightCountAsync(
```

Comment:
Would it be reasonable to move all of the lock holding out to here?  This method immediately enters the lock, as does `solution.GetSolutionAsync` in the try and  `solution.DecrementInFlightCount();` in the finally.

It it slightly difficult to reason about the order when they're all nested inside the helpers.

Reply:
yes.  that made it much simpler.